socs/agents/hwp_pid/drivers/pid_controller.py

```diff
@@ -0,0 +1,351 @@
+
+    def set_scale(self, slope, offset):
+        """Set the conversion between feedback voltage and approximate
+        frequency.
+
+        """
+        slope_hex = self._get_scale_hex(slope, 1)
+        offset_hex = self._get_scale_hex(offset, 2)
+
+        responses = []
+        responses.append(self.send_message(f"*W14{slope_hex}"))
+        responses.append(self.send_message(f"*W03{offset_hex}"))
+        responses.append(self.send_message("*Z02"))
+        if self.verb:
+            print(responses)
+            print(self.return_messages(responses))
+
+    ######################################################################
+    # Messaging
+    ######################################################################
+
+    def send_message(self, msg):
+        """Send message over TCP to the PID controller.
+
+        Args:
+            msg (str): Command to send to the controller.
+
+        Returns:
+            str: Respnose from the controller.
+
+        """
+        self.conn.sendall((msg + '\r\n').encode())
+        time.sleep(0.5)  # Don't send messages too quickly
+        for attempt in range(2):
+            try:
+                data = self.conn.recv(4096).decode().strip()
+                break
+            except socket.timeout:
+                print("Caught timeout waiting for response from PID controller. "
+                      + "Trying again...")
+                time.sleep(1)
+                if attempt == 1:
+                    raise RuntimeError(
+                        'Response from PID controller timed out.')
+        return data
+
+    def return_messages(self, msg):
+        """Decode list of responses from PID controller and return useful
+        values.
+
+        Args:
+            msg (list): List of messages to decode.
+
+        Returns:
+            list: Decoded responses.
+
+        """
+        return self._decode_array(msg)
+
+    @staticmethod
+    def _decode_array(input_array):
+        output_array = list(input_array)
+
+        for index, string in enumerate(list(input_array)):
+            header = string[0]
+
+            if header == 'R':
+                output_array[index] = PID._decode_read(string)
+            elif header == 'W':
+                output_array[index] = PID._decode_write(string)
+            elif header == 'E':
+                output_array[index] = 'PID Enabled'
+            elif header == 'D':
+                output_array[index] = 'PID Disabled'
+            elif header == 'P':
+                pass
+            elif header == 'G':
+                pass
+            elif header == 'X':
+                output_array[index] = PID._decode_measure(string)
+            else:
+                pass
+
+        return output_array
+
+    def _decode_read(self, string):
+        read_type = string[1:3]
+        if read_type == '01':
+            self.target = float(int(string[4:], 16) / 1000.)
+            return 'Setpoint = ' + str(self.target)
```

Comment:
I don't think `@staticmethod` should be removed from this. The value that you're setting `self.target` to gets returned and passed up through to `return_messages()`, `self.target` should be set where you call `return_messages()` (like in `set_target` or `get_target` or both), not all the way down here.

I removed a similar structure from the direction reading back here: https://github.com/simonsobs/socs/commit/634ec8506a136aec6e50e21e2fd7b27346f1f8bf

Reply:
Thanks for the suggestion. I fixed the structure of target freq reading. https://github.com/simonsobs/socs/pull/421/commits/3a9d36da49c1b3626c07f419cc8526fbe947e4fe